src/adapter.rs

```diff
@@ -28,30 +30,43 @@ use crate::IntrusivePointer;
 /// given type and its link field. However it is possible to implement it
 /// manually if the intrusive link is not a direct field of the object type.
 ///
-/// It is also possible to create stateful adapters. This allows links and
-/// containers to be separated and avoids the need for objects to be modified to
+/// It is also possible to create stateful adapters.
+/// This allows links and containers to be separated and avoids the need for objects to be modified to
 /// contain a link.
 ///
 /// # Safety
 ///
 /// It must be possible to get back a reference to the container by passing a
 /// pointer returned by `get_link` to `get_container`.
 pub unsafe trait Adapter {
-    /// Collection-specific link type which allows an object to be inserted in
+    /// Collection-specific link operations which allow an object to be inserted in
     /// an intrusive collection.
-    type Link;
+    type LinkOps: LinkOps;
 
-    /// Object type which is inserted in an intrusive collection.
-    type Value: ?Sized;
-
-    /// Pointer type which owns an instance of a value.
-    type Pointer: IntrusivePointer<Self::Value>;
+    /// Collection-specific pointer conversions which allow an object to
+    /// be inserted in an intrusive collection.
+    type PointerOps: PointerOps;
 
     /// Gets a reference to an object from a reference to a link in that object.
-    unsafe fn get_value(&self, link: *const Self::Link) -> *const Self::Value;
+    unsafe fn get_value(
+        &self,
+        link: <Self::LinkOps as LinkOps>::LinkPtr,
+    ) -> *const <Self::PointerOps as PointerOps>::Value;
 
     /// Gets a reference to the link for the given object.
-    unsafe fn get_link(&self, value: *const Self::Value) -> *const Self::Link;
+    unsafe fn get_link(
+        &self,
+        value: *const <Self::PointerOps as PointerOps>::Value,
+    ) -> <Self::LinkOps as LinkOps>::LinkPtr;
+
+    /// Returns a reference to the link operations.
+    fn link_ops(&self) -> &Self::LinkOps;
+
+    /// Returns a reference to the mutable link operations.
+    fn link_ops_mut(&mut self) -> &mut Self::LinkOps;
+
+    /// Returns a reference to the pointer converter.
+    fn pointer_ops(&self) -> &Self::PointerOps;
```

Comment:
This breaks `Cursor::clone_pointer` since it requires a non-mutable reference to the underlying intrusive collection.

```rust
impl<'a, A> Cursor<'a, A>
where
    A::LinkOps: LinkedListOps
{
    /// Clones and returns the pointer that points to the element that the
    /// cursor is referencing.
    ///
    /// This returns `None` if the cursor is currently pointing to the null
    /// object.
    #[inline]
    pub fn clone_pointer(&self) -> Option<<A::PointerOps as PointerOps>::Pointer>
    where
        <A::PointerOps as PointerOps>::Pointer: Clone,
    {
        let raw_pointer = self.get()? as *const <A::PointerOps as PointerOps>::Value;
        Some(unsafe {
            crate::pointer_ops::clone_pointer_from_raw(self.list.adapter.pointer_ops(), raw_pointer)
        })
    }
}
```